pkg/volume/glusterfs/glusterfs.go

```diff
@@ -0,0 +1,233 @@
+		volName: spec.Name,
+		hosts:   spec.VolumeSource.Glusterfs.Hosts,
+		path:    spec.VolumeSource.Glusterfs.Path,
+		option:  spec.VolumeSource.Glusterfs.MountOpt,
+		helper:  spec.VolumeSource.Glusterfs.Helper,
+		mounter: mounter,
+		exe:     exe,
+		podRef:  podRef,
+		plugin:  plugin,
+	}, nil
+}
+
+func (plugin *glusterfsPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, mount.New())
+}
+
+func (plugin *glusterfsPlugin) newCleanerInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &glusterfs{
+		volName: volName,
+		mounter: mounter,
+		podRef:  &api.ObjectReference{UID: podUID},
+		plugin:  plugin,
+	}, nil
+}
+
+// Glusterfs volumes represent a bare host file or directory mount of an Glusterfs export.
+type glusterfs struct {
+	volName string
+	podRef  *api.ObjectReference
+	hosts   []string
+	path    string
+	option  string
+	helper  string
+	mounter mount.Interface
+	exe     exec.Interface
+	plugin  *glusterfsPlugin
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (glusterfsVolume *glusterfs) SetUp() error {
+	return glusterfsVolume.SetUpAt(glusterfsVolume.GetPath())
+}
+
+func (glusterfsVolume *glusterfs) SetUpAt(dir string) error {
+	mountpoint, err := mount.IsMountPoint(dir)
+	glog.V(4).Infof("Glusterfs: mount set up: %s %v %v", dir, mountpoint, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if mountpoint {
+		return nil
+	}
+	path := glusterfsVolume.path
+	os.MkdirAll(dir, 0750)
+	err = glusterfsVolume.execMount(glusterfsVolume.hosts, path, dir, glusterfsVolume.option, glusterfsVolume.helper)
+	if err != nil {
+		mountpoint, mntErr := mount.IsMountPoint(dir)
+		if mntErr != nil {
+			glog.Errorf("Glusterfs: IsMountpoint check failed: %v", mntErr)
+			return err
+		}
+		if mountpoint {
+			if mntErr = glusterfsVolume.mounter.Unmount(dir, 0); mntErr != nil {
+				glog.Errorf("Glusterfs: Failed to unmount: %v", mntErr)
+				return err
+			}
+			mountpoint, mntErr := mount.IsMountPoint(dir)
+			if mntErr != nil {
+				glog.Errorf("Glusterfs: IsMountpoint check failed: %v", mntErr)
+				return err
+			}
+			if mountpoint {
+				glog.Errorf("Glusterfs: cannot unmount %s.", dir)
+				return err
+			}
+		}
+		os.Remove(dir)
+		return err
+	}
+	return nil
+}
+
+func (glusterfsVolume *glusterfs) GetPath() string {
+	name := glusterfsPluginName
+	return glusterfsVolume.plugin.host.GetPodVolumeDir(glusterfsVolume.podRef.UID, util.EscapeQualifiedNameForDisk(name), glusterfsVolume.volName)
+}
+
+func (glusterfsVolume *glusterfs) TearDown() error {
+	return glusterfsVolume.TearDownAt(glusterfsVolume.GetPath())
+}
+
+func (glusterfsVolume *glusterfs) TearDownAt(dir string) error {
+	mountpoint, err := mount.IsMountPoint(dir)
+	if err != nil {
+		glog.Errorf("Glusterfs: Error checking IsMountPoint: %v", err)
+		return err
+	}
+	if !mountpoint {
+		return os.Remove(dir)
```

Comment:
Suggestion: Use `os.RemoveAll` since it will not return error if `dir` is not found. This helps make TearDown idempotent.